Finish splitting functionality of the `RenderingDevice` backends into `RenderingDeviceDriver`.

DarioSamo opened the pull request:
This is a follow up PR to @RandomShaper's PR https://github.com/godotengine/godot/pull/83452 with the intention of cleaning up the loose ends that weren't put behind an abstraction yet. This also includes the changes identified by @darksylinc's PR https://github.com/godotengine/godot/pull/80566 as well as the new project settings that allow configuration of the swap chain image count and the buffered frame count of RenderingDevice.

The main reason behind pushing for this change is that the additional flexibility provided by the PR is essential to providing new asynchronous transfer queues, which will lead to [massive improvements in loading times](https://twitter.com/dariosamo/status/1735363059202375819) (not included in this PR). This will also enable to provide additional multi-queue functionality in the future for the Acyclic Command Graph that was recently merged. However, the current design makes it impossible to implement this without making awkward changes to all backends and maintaining all of the backends at once.

Another important reason for this change is that this PR will finally enable the possibility of the RenderingDevice's local device working independently of the renderer that was chosen, **which means the GPU lightmapper can finally work in Compatibility mode!** @clayjohn has proven this to be possible in a separate experiment (not part of this PR but based on top of it) where he added the required UV2 rendering to the compatibility renderer.

The overall net result means a lot of the logic is now shared across backends and the only thing the backend needs to worry about is exposing the following functionality:
- Command Queues
- CPU/GPU Fences
- GPU/GPU Semaphores
- Swap Chains

The resulting code is also a lot easier to follow and it should be much simpler to fix some of the [strange bugs that have been plaguing](https://github.com/godotengine/godot/issues/71929) the backends for a while. However, I do not expect my first attempt at this to have been perfect, since swap chain logic can be particularly hard to handle correctly, so we'll need some extensive testing to confirm that this doesn't introduce any regressions.

NOTE: No performance improvements or regressions are expected from this PR, it is purely an organizational change and functionality wise it should be identical in the default case.

Confirm the changes work as intended in all the affected platforms.
- [x] Windows
- [x] Linux (X11 Systems)
- [x] Linux (Wayland)
- [x] Android
- [x] macOS
- [x] iOS
- [x] OpenXR (Desktop)
- [x] OpenXR (Quest)

TODO:
- [x] CmdBeginDebugUtilsLabelEXT and related functions should be fetched like the older versions.
- [x] Rendering Device should print the API name instead of "API".
- [x] There should be no need to print the information about the picked device for RDs that aren't the Singleton.
- [x] Fix all CI issues so it builds successfully in all platforms.
- [x] Provide compatibility methods if necessary for RenderingDevice on the functions that have been changed.
- [x] Finish documentation changes.

___

- *Production edit: This closes https://github.com/godotengine/godot/issues/87532.*

DarioSamo committed:
Finish splitting functionality of the Vulkan and D3D12 backends into RenderingDeviceDriver.

AThousandShips commented on doc/classes/ProjectSettings.xml:
```diff
@@ -2656,6 +2656,21 @@
+		<member name="rendering/rendering_device/vsync/swapchain_image_count" type="int" setter="" getter="" default="3">
+			The number of images the swapchain will consist of (back buffers + front buffer).
+			[code]2[/code] corresponds to double-buffering and [code]3[/code] to triple-buffering. A value of [code]1[/code] is not recommended.
+			Double-buffering may give you the lowest lag/latency but if V-Sync is on and the system can't render at 60 fps, the framerate will go down in multiples of it (e.g. 30 fps, 15, 7.5, etc ). Triple buffering gives you higher framerate (specially if the system can't reach a constant 60 fps) at the cost of up to 1 frame of latency, with [constant DisplayServer.VSYNC_ENABLED] (FIFO).
```
```suggestion
			Double-buffering may give you the lowest lag/latency but if V-Sync is on and the system can't render at 60 fps, the framerate will go down in multiples of it (e.g. 30 fps, 15, 7.5, etc.). Triple buffering gives you higher framerate (specially if the system can't reach a constant 60 fps) at the cost of up to 1 frame of latency, with [constant DisplayServer.VSYNC_ENABLED] (FIFO).
```

AThousandShips commented on drivers/d3d12/rendering_context_driver_d3d12.cpp:
```diff
@@ -0,0 +1,313 @@
+		}
+	} while (adapter != nullptr);
+
+	ERR_FAIL_COND_V_MSG(adapters.size() == 0, ERR_CANT_CREATE, "Adapters enumeration reported zero accessible devices.");
```
```suggestion
	ERR_FAIL_COND_V_MSG(adapters.is_empty(), ERR_CANT_CREATE, "Adapters enumeration reported zero accessible devices.");
```

AThousandShips commented on drivers/d3d12/rendering_device_driver_d3d12.cpp:
```diff
@@ -1772,25 +1861,173 @@ void RenderingDeviceDriverD3D12::command_pipeline_barrier(
+	ERR_FAIL_COND_V(!SUCCEEDED(res), FenceID());
+
+	HANDLE event_handle = CreateEvent(nullptr, FALSE, FALSE, nullptr);
+	ERR_FAIL_COND_V(event_handle == nullptr, FenceID());
```
```suggestion
	ERR_FAIL_NULL_V(event_handle, FenceID());
```

AThousandShips commented on drivers/d3d12/rendering_device_driver_d3d12.cpp:
```diff
@@ -5408,109 +5828,394 @@ uint64_t RenderingDeviceDriverD3D12::api_trait_get(ApiTrait p_trait) {
+Error RenderingDeviceDriverD3D12::initialize(uint32_t p_device_index, uint32_t p_frame_count) {
+	context_device = context_driver->device_get(p_device_index);
+	adapter = context_driver->create_adapter(p_device_index);
+	ERR_FAIL_COND_V(adapter == nullptr, ERR_CANT_CREATE);
```
```suggestion
	ERR_FAIL_NULL_V(adapter, ERR_CANT_CREATE);
```

AThousandShips commented on drivers/vulkan/rendering_context_driver_vulkan.cpp:
```diff
@@ -0,0 +1,690 @@
+					" - " + string_VkObjectType(p_callback_data->pObjects[object].objectType) +
+					", Handle " + String::num_int64(p_callback_data->pObjects[object].objectHandle);
+
+			if (nullptr != p_callback_data->pObjects[object].pObjectName && strlen(p_callback_data->pObjects[object].pObjectName) > 0) {
```
```suggestion
			if (p_callback_data->pObjects[object].pObjectName != nullptr && strlen(p_callback_data->pObjects[object].pObjectName) > 0) {
```

AThousandShips commented on modules/lightmapper_rd/lightmapper_rd.cpp:
```diff
@@ -1017,7 +1021,37 @@ LightmapperRD::BakeError LightmapperRD::bake(BakeQuality p_quality, bool p_use_d
+	}
+
+	if (rd == nullptr) {
+		ERR_FAIL_V(BAKE_ERROR_LIGHTMAP_CANT_PRE_BAKE_MESHES);
```
This should be a `ERR_FAIL_NULL_V`

darksylinc commented:
I've been reviewing it today, the Vulkan & generic parts look alright (I did not review D3D12).

As I already told Dario personally, the only thing that makes me nervous is this:

```
RenderingDeviceGraph::frame
RenderingDevice::frame
```

They're both cloned versions and there's no guarantee(*) they're in sync. But on the other hand, if we'd unify them into just `RenderingDevice`, we still have no guarantee `RenderingDeviceGraph` will notice when `RenderingDevice::frame` advances.

And what's missing would be to aggressively check:
```
DEV_ASSERT( RenderingDevice::frame == RenderingDeviceGraph::frame );
```

Nonetheless, as Dario explained to me and I've confirmed, `RenderingDeviceGraph::frame` isn't currently in use because that's only used for secondary command buffers, which had to be disabled because they were broken on NVIDIA (either because of NV errors, or because of Godot's faults).


(*) By "no guarantee" I mean if someone in the future touches this code and inadvertently introduces subtle bugs. Even if the current code is 100% bug free; checks need to be done to ensure it doesn't get broken tomorrow.

BastiaanOlij commented:
Haven't had time to dive deep into this but making sure OpenXR still works I can confirm that:
- PCVR works which suggests our OpenXR overrules are still working
- Running on Quest this currently hard crashes but I haven't been able to capture good crash info. Needs to be tested whether this is an Android issue or specific to Android XR devices.

DarioSamo commented:
It was indeed crashing in Android due to an implementation error on my part. Now I've checked and it seems to be working. Weirdly enough, it also seems to fix a bug where using the controls in the editor would stretch the contents of the viewports to be twice its scale.

Calinou reviewed (approved):
Tested locally on Linux + NVIDIA, it works as expected.

However, while this resolves https://github.com/godotengine/godot/issues/87532, every time I bake lightmaps, I now see this message in the Output panel:

```
API 1.3.260 - Forward+ - Using Device #0: NVIDIA - NVIDIA GeForce RTX 4090
```

It's missing "Vulkan" before "API", and I think there should be a way to silence it as it may be excessive for certain local RenderingDevices.

DarioSamo commented:
> However, while this resolves #87532, every time I bake lightmaps, I now see this message in the Output panel:
> 
> ```
> API 1.3.260 - Forward+ - Using Device #0: NVIDIA - NVIDIA GeForce RTX 4090
> ```
> 
> It's missing "Vulkan" before "API", and I think there should be a way to silence it as it may be excessive for certain local RenderingDevices.

Noted, I'll implement these suggestions.

reduz commented:
Fantastic work on this!!

DarioSamo commented:
Opening this up for review, I'll need some help in testing out the other platforms remaining in the TODO list as I don't own the hardware to verify this.

AThousandShips reviewed (approved):
Some style details but looks good otherwise for style and macro usage

DarioSamo commented on drivers/d3d12/rendering_context_driver_d3d12.h:
```diff
@@ -0,0 +1,120 @@
+	void surface_set_needs_resize(SurfaceID p_surface, bool p_needs_resize) override final;
+	bool surface_get_needs_resize(SurfaceID p_surface) const override final;
+	void surface_destroy(SurfaceID p_surface) override final;
+	bool is_debug_utils_enabled() const override final;
```
Fair enough, also taking out some of these finals as well. They were mostly copied from the other style of files but it'll be quite problematic in the future for a platform that just needs to inherit and change a few things of these drivers.

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -2658,6 +2658,21 @@
 		</member>
+		<member name="rendering/rendering_device/vsync/frame_queue_size" type="int" setter="" getter="" default="2">
+			The number of frames to track on the CPU side before stalling to wait for the GPU.
+			1 may give you the lowest lag/latency but at the high cost of no parallelism between CPU &amp; GPU.
```
Could also start with "A value of", but honestly the subject is fairly clear here.
The ampersand is used extremely rarely in the class reference, too, but I wouldn't split hairs over it, if you feel like it looks better.
```suggestion
			[code]1[/code] may give you the lowest lag/latency but at the high cost of no parallelism between CPU and GPU.
```

DarioSamo replied:
Same reason as the other value, these are clamped to 2 as well at minimum to avoid introducing "footguns" the users could misconfigure.

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -2658,6 +2658,21 @@
+		</member>
+		<member name="rendering/rendering_device/vsync/swapchain_image_count" type="int" setter="" getter="" default="3">
+			The number of images the swapchain will consist of (back buffers + front buffer).
+			[code]2[/code] corresponds to double-buffering and [code]3[/code] to triple-buffering. A value of [code]1[/code] is not recommended.
```
> A value of [code]1[/code] is not recommended.

Why? If it can be summarised with "It just does not look good" feel free to mention it. If the explanation is too long I'm okay with omitting it.

darksylinc replied:
The whole block of text has gone through multiple revisions.

The problem is that this setting is related to the following 3 topics:

 - V-Sync
 - Input to Display Latency
 - CPU & GPU synchronization

One could write a 100-page (or more) essay of these 3. There is only so much that can be squashed into a tooltip. They are all advanced topics.

This is why the tooltip refers to the [Simulator](https://darksylinc.github.io/vsync_simulator) which contains an interactive simulator and [links](https://askubuntu.com/a/1420575/274245) to more [bibliography](https://www.gdcvault.com/play/1026327/Controller-to-Display-Latency-in) about it that go in depth.

Regarding this particular question. A value of 1is not recommended, is because the CPU will always have to wait for the GPU, which can hurt performance a lot, but might be important for specific scenarios where latency must be the lowest (real time medical imagining _might_ need this), or RAM is tight and synchronization is not necessary or rare (AI and blockchain mining comes to mind).

However I saw that this PR from @DarioSamo always performs a `actual_count = max( 2u, swapchain_image_count )` unless local device is used. I don't know exactly why he did this, but it may make this comment irrelevant and maybe `swapchain_image_count` should have 2 as min value.

DarioSamo replied:
We consulted with @reduz and we should just have two as the safe minimum, so the note will be pretty irrelevant.

Mickeon commented on doc/classes/ProjectSettings.xml:
```diff
@@ -2658,6 +2658,21 @@
+			Use double-buffering with [constant DisplayServer.VSYNC_ENABLED]. Triple-buffering is a must if you plan on using [constant DisplayServer.VSYNC_MAILBOX] mode.
+			Try the [url=https://darksylinc.github.io/vsync_simulator/]V-Sync Simulator[/url], an interactive interface that simulates presentation to better understand how it is affected by different variables under various conditions.
+			[b]Note:[/b] This property is only read when the project starts. There is currently no way to change this value at run-time.
+			[b]Note:[/b] Some platforms may restrict the actual value.
```
This note is a bit too up to interpretation. It would be lovely to mention a few platforms that do this, by name.

darksylinc replied:
Since the platforms and API backends are evolving over time, it's hard to say.

I know for sure these values are ignored in OpenGL (compatibility). But the other backends (D3D12, Metal) are not finished yet. It's also unclear what happens in macOS via MoltenVK.

Mickeon replied:
In that case, if the compatibility renderer is guaranteed to stay like this, that's worth mentioning (it also encompasses other platforms like Web).

DarioSamo replied:
Rendering Device is not used by the compatibility renderer, which is where this setting lives on in the configuration panel. It doesn't sound like mentioning Compatibility renderer is relevant in that case to me.

The actual values that might be restricted by platforms is a pretty mixed bag to give clear examples and seems to change every now and then. I'd suggest a note that just says the value is a suggestion to the OS but that it might not match what the user actually requests at the low level.

Mickeon reviewed (commented):
Some thoughts about the class reference (which I like a lot by the way)

DarioSamo commented:
@Mickeon Those were taken from @darksylinc's other PR with his permission so I'll have to refer to him to see if he agrees with the changes, as I did not write these originally.

HybridEidolon commented:
As currently written, this seems to partially fix #85547. The shadow flickering issue now only happens when:

- Resizing the window
- Holding Alt for the Alt-Tab widget

Maximizing and minimizing on W11 with the AMD driver work great. Resizing works fine with aformentioned issue.

DarioSamo commented:
Rebased to integrate the latest changes to D3D12.

Still looking for confirmation for Apple devices from anyone who owns them!

HybridEidolon commented:
@DarioSamo I built it on my M1 Macbook Air (only for x86_64, for some reason arm64 isn't able to find mvk symbols) and it works fine in a basic scene with SDFGI enabled. Resizing, moving, fullscreen, all of that works great. Just shader compilation slowness.

Calinou commented:
> Edit: Spoke too soon; minimizing to tray does appear to cause issues with re-renders, but I'm not sure if that's new to this branch or something that's always existed on macOS.

This is likely https://github.com/godotengine/godot/issues/87309.

DarioSamo commented:
> Edit: Spoke too soon; minimizing to tray does appear to cause issues with re-renders, but I'm not sure if that's new to this branch or something that's always existed on macOS.

If you can confirm it's not a regression from `master` then that's probably fine.

BastiaanOlij commented:
Still having problems running this on Quest:
![image](https://github.com/godotengine/godot/assets/1945449/ce0abd00-ef75-49e9-8237-4921bebe420f)

It looks like the injection of OpenXR code to initialise Vulkan is not happening correctly

DarioSamo commented:
Thanks to @BastiaanOlij we tracked down that it is not necessary to request the transfer bit for the main queue, as apparently it is optional to do so and the Quest does not define it despite requesting the graphics and compute bits. This confirms that OpenXR does work in Quest and so all of the platforms we were looking to verify are working.

BastiaanOlij commented:
> Thanks to @BastiaanOlij we tracked down that it is not necessary to request the transfer bit for the main queue, as apparently it is optional to do so and the Quest does not define it despite requesting the graphics and compute bits. This confirms that OpenXR does work in Quest and so all of the platforms we were looking to verify are working.

Such a weird problem, but glad we figured it out. This all works great imho

RandomShaper commented on servers/rendering/rendering_device_graph.cpp:
```diff
@@ -40,15 +40,7 @@ RenderingDeviceGraph::RenderingDeviceGraph() {
-			}
-		}
-	}
+	// Default destruction.
```
```suggestion
```

RandomShaper commented on drivers/d3d12/rendering_context_driver_d3d12.cpp:
```diff
@@ -0,0 +1,314 @@
+}
+
+RenderingContextDriverD3D12::~RenderingContextDriverD3D12() {
+	// Does nothing.
```
```suggestion
```

RandomShaper commented on drivers/d3d12/rendering_context_driver_d3d12.cpp:
```diff
@@ -0,0 +1,314 @@
+
+Error RenderingContextDriverD3D12::_initialize_debug_layers() {
+	ComPtr<ID3D12Debug> debug_controller;
+	HRESULT res;
```
```suggestion
	HRESULT res = OK;
```

DarioSamo replied:
What's the point of the default initialization if it's pretty much guaranteed it'll have a value from either branch of the code? The sanitization can easily detect something went wrong if the code ahead is affected, while a default initialization might hide the issue. That's pretty much the reasoning in case you were wondering.

DarioSamo replied:
Should we mark this as resolved?

RandomShaper commented on drivers/d3d12/rendering_context_driver_d3d12.h:
```diff
@@ -0,0 +1,120 @@
+#define ARRAY_SIZE(a) (sizeof(a) / sizeof(a[0]))
+
+class RenderingContextDriverD3D12 : public RenderingContextDriver {
+private:
```
```suggestion
```

RandomShaper commented on drivers/d3d12/rendering_context_driver_d3d12.h:
```diff
@@ -0,0 +1,120 @@
+	Error _initialize_devices();
+
+public:
+	RenderingContextDriverD3D12();
```
I'm too late for `RenderingDeviceGraph`, but Godot codebase has a standard of putting constructors and destructors as the last members of classes. (This also applies to other offending cases in this PR.)

DarioSamo replied:
Done.

RandomShaper commented on drivers/d3d12/rendering_context_driver_d3d12.h:
```diff
@@ -0,0 +1,120 @@
+
+	// Platform-specific data for the Windows embedded in this driver.
+	struct WindowPlatformData {
+		HWND window;
```
```suggestion
	HWND window = nullpr;
```

(Well, or `NULL`, as you did below, which is more Win32-ish, although in Godot `nullptr` is generally favored even when interacting with C APIs.)

DarioSamo replied:
I'd stick with NULL if using Win32.

DarioSamo replied:
I just remembered we can't do this. This is part of a union, we can't have default initialization on the struct.

RandomShaper replied:
Oh, indeed.

RandomShaper commented on drivers/d3d12/rendering_device_driver_d3d12.h:
```diff
@@ -64,15 +64,13 @@ using Microsoft::WRL::ComPtr;
 
 #define D3D12_BITCODE_OFFSETS_NUM_STAGES 3
 
+// Forward declarations.
```
```suggestion
```

RandomShaper commented on drivers/d3d12/rendering_device_driver_d3d12.h:
```diff
@@ -86,8 +84,58 @@ class RenderingDeviceDriverD3D12 : public RenderingDeviceDriver {
+	};
+
+	struct SubgroupCapabilities {
+		uint32_t size;
+		bool wave_ops_supported;
```
```suggestion
		uint32_t size = 0;
		bool wave_ops_supported = false;
```

RandomShaper commented on drivers/d3d12/rendering_device_driver_d3d12.h:
```diff
@@ -86,8 +84,58 @@ class RenderingDeviceDriverD3D12 : public RenderingDeviceDriver {
+	};
+
+	struct VRSCapabilities {
+		bool draw_call_supported; // We can specify our fragment rate on a draw call level.
+		bool primitive_supported; // We can specify our fragment rate on each drawcall.
+		bool primitive_in_multiviewport;
+		bool ss_image_supported; // We can provide a density map attachment on our framebuffer.
+		uint32_t ss_image_tile_size;
+		bool additional_rates_supported;
```
```suggestion
		bool draw_call_supported = false; // We can specify our fragment rate on a draw call level.
		bool primitive_supported = false; // We can specify our fragment rate on each drawcall.
		bool primitive_in_multiviewport = false;
		bool ss_image_supported = false; // We can provide a density map attachment on our framebuffer.
		uint32_t ss_image_tile_size = 0;
		bool additional_rates_supported = false;
```

RandomShaper commented on drivers/d3d12/rendering_device_driver_d3d12.h:
```diff
@@ -86,8 +84,58 @@ class RenderingDeviceDriverD3D12 : public RenderingDeviceDriver {
+	};
+
+	struct ShaderCapabilities {
+		D3D_SHADER_MODEL shader_model;
+		bool native_16bit_ops;
```
```suggestion
		D3D_SHADER_MODEL shader_model = (D3D_SHADER_MODEL)0;
		bool native_16bit_ops = false;
```

RandomShaper commented on drivers/d3d12/rendering_device_driver_d3d12.h:
```diff
@@ -86,8 +84,58 @@ class RenderingDeviceDriverD3D12 : public RenderingDeviceDriver {
+	};
+
+	struct StorageBufferCapabilities {
+		bool storage_buffer_16_bit_access_is_supported;
```
```suggestion
		bool storage_buffer_16_bit_access_is_supported = false;
```

RandomShaper commented on drivers/d3d12/rendering_device_driver_d3d12.h:
```diff
@@ -86,8 +84,58 @@ class RenderingDeviceDriverD3D12 : public RenderingDeviceDriver {
+	};
+
+	struct FormatCapabilities {
+		bool relaxed_casting_supported;
```
```suggestion
		bool relaxed_casting_supported = false;
```

RandomShaper commented on servers/rendering/rendering_context_driver.cpp:
```diff
@@ -0,0 +1,86 @@
+#include "rendering_context_driver.h"
+
+RenderingContextDriver::~RenderingContextDriver() {
+	// Does nothing.
```
```suggestion
```

RandomShaper reviewed (changes requested):
This is in such a good level of quality that I haven't been able but to insult it with nitpickery. 😀

Mickeon commented on doc/classes/RenderingDevice.xml:
```diff
@@ -590,6 +590,7 @@
 			<return type="int" />
+			<param index="0" name="screen" type="int" default="0" />
 			<description>
 				Returns the screen's framebuffer format.
```
Since you added a screen parameter it may be worth being more explicit.
```suggestion
				Returns the framebuffer format of the given screen.
```

DarioSamo replied:
Agreed.

DarioSamo commented:
Seems like I'll have to re-do my changes for Wayland, testing would be appreciated when that is done as well.

DarioSamo commented:
> Oof some recent merges just conflicted this PR. I'm still testing this, but this already needs a rebase xD

You mean like in the middle of the hour I was just working on this? 😱 

Let me re-do it.

DarioSamo commented:
@Riteo Alright, conflicts should be fixed now.

DarioSamo commented:
> I tried this branch briefly with the Wayland backend on my crappy laptop and I couldn't notice any particular artifacts on the [Garage 42 demo](https://godotengine.org/asset-library/asset/2379) (I'll test properly on my desktop later).
> 
> There's a segfault on exit though:

I think that's a good enough confirmation for me TBH, the segfault sounds unrelated to the platform and something I might not just be cleaning up properly, so I'll review it. Thanks for the test!

DarioSamo commented:
> There's a segfault on exit though:
> 
> ```
> Thread 1 "godot.linuxbsd." received signal SIGILL, Illegal instruction.
> 0x000055555f098bae in LocalVector<RenderingDevice::Frame, unsigned int, false, true>::operator[] (this=0x7fffef3aac68, p_index=1) at ./core/templates/local_vector.h:161
> 161			CRASH_BAD_UNSIGNED_INDEX(p_index, count);
> (gdb) bt
> #0  0x000055555f098bae in LocalVector<RenderingDevice::Frame, unsigned int, false, true>::operator[](unsigned int) (this=0x7fffef3aac68, p_index=1) at ./core/templates/local_vector.h:161
> #1  0x000055555efb3265 in RenderingDevice::_end_frame() (this=0x7fffef3aa020) at ./servers/rendering/rendering_device.cpp:4887
> #2  0x000055555efb3a6c in RenderingDevice::_flush_and_stall_for_all_frames() (this=0x7fffef3aa020) at ./servers/rendering/rendering_device.cpp:4919
> #3  0x000055555efb70e7 in RenderingDevice::finalize() (this=0x7fffef3aa020) at ./servers/rendering/rendering_device.cpp:5323
> #4  0x000055555efd5746 in RenderingDevice::~RenderingDevice() (this=0x7fffef3aa020, __in_chrg=<optimized out>) at ./servers/rendering/rendering_device.cpp:6123
> #5  0x000055555afc4b34 in memdelete<RenderingDevice>(RenderingDevice*) (p_class=0x7fffef3aa020) at ./core/os/memory.h:109
> #6  0x000055555aff72ee in DisplayServerWayland::~DisplayServerWayland() (this=0x7ffff716b9a0, __in_chrg=<optimized out>) at platform/linuxbsd/wayland/display_server_wayland.cpp:1386
> #7  0x000055555aff923c in memdelete<DisplayServer>(DisplayServer*) (p_class=0x7ffff716b9a0) at ./core/os/memory.h:109
> #8  0x000055555b0257bc in finalize_display() () at main/main.cpp:325
> #9  0x000055555b04232f in Main::cleanup(bool) (p_force=false) at main/main.cpp:4056
> #10 0x000055555af802da in main(int, char**) (argc=3, argv=0x7fffffffe2e8) at platform/linuxbsd/godot_linuxbsd.cpp:76
> ```

Turns out this is actually related to the platform, but only because I added a call to "finalize()" by mistake twice to RenderingDevice when no other platform was doing it. My bad. Either way I've fixed RD to be resilient against finalizing twice and also made it so it's not called twice on Wayland.

clayjohn reviewed (approved):
I've reviewed the code and tested locally. This PR has also seen extensive revision and testing from others. 

I am satisfied at this point that it is ready to merge

It just needs a quick rebase

DarioSamo commented:
Rebased to fixed conflicts.

bruvzg commented on drivers/d3d12/rendering_device_driver_d3d12.cpp:
```diff
@@ -1858,22 +2090,219 @@ void RenderingDeviceDriverD3D12::command_buffer_end(CommandBufferID p_cmd_buffer
+		texture_info.owner_info.states.subresource_states.push_back(D3D12_RESOURCE_STATE_PRESENT);
+		texture_info.states_ptr = &texture_info.owner_info.states;
+		texture_info.format = swap_chain->data_format;
+		texture_info.desc = CD3DX12_RESOURCE_DESC(render_target->GetDesc());
```
This breaks MinGW build.

DarioSamo replied:
I checked your other PR, how strange that the headers literally define it differently. Now it's a bit late but maybe we have to make a wrapper around that function that works for both compilers without putting the compiler paths inside the caller methods.

bruvzg commented on drivers/d3d12/rendering_device_driver_d3d12.cpp:
```diff
@@ -5224,17 +5652,9 @@ void RenderingDeviceDriverD3D12::begin_segment(CommandBufferID p_cmd_buffer, uin
 	frames[frame_idx].desc_heap_walkers.rtv.rewind();
 	frames[frame_idx].desc_heaps_exhausted_reported = {};
-	frames[frame_idx].null_rtv_handle = { 0 };
+	frames[frame_idx].null_rtv_handle = {};
```
Also, breaks MinGW build.

bruvzg commented:
Seems like some change in this PR broke multi-threaded renderer (both compatibility and Vulkan) on macOS (see https://github.com/godotengine/godot/pull/90268#discussion_r1554267653), but I'm not what's exactly wrong.

Alex2782 commented:
**RenderPass2KHR** - `Compatibility fallback` does not work properly on Android LG Nexus 5X, Adreno 418

Also same crash on MacOS without:
``` cpp
_register_requested_device_extension(VK_KHR_CREATE_RENDERPASS_2_EXTENSION_NAME, false);
```



--------------------------

https://github.com/godotengine/godot/blob/780e1a50408360cf0cf93c0b55b59e9d7b4ad0b1/drivers/vulkan/rendering_device_driver_vulkan.cpp#L979

### **crash at:**
https://github.com/godotengine/godot/blob/780e1a50408360cf0cf93c0b55b59e9d7b4ad0b1/drivers/vulkan/rendering_device_driver_vulkan.cpp#L1011

https://github.com/godotengine/godot/blob/780e1a50408360cf0cf93c0b55b59e9d7b4ad0b1/drivers/vulkan/rendering_device_driver_vulkan.cpp#L970

-------------------------------
If I **deactivate the extension** (`VK_KHR_CREATE_RENDERPASS_2_EXTENSION_NAME`), I can also reproduce the crash on **MacMini M1.**

https://github.com/godotengine/godot/blob/780e1a50408360cf0cf93c0b55b59e9d7b4ad0b1/drivers/vulkan/rendering_device_driver_vulkan.cpp#L435
```cpp
Error RenderingDeviceDriverVulkan::_initialize_device_extensions() {
	enabled_device_extension_names.clear();

	_register_requested_device_extension(VK_KHR_SWAPCHAIN_EXTENSION_NAME, true);
	_register_requested_device_extension(VK_KHR_MULTIVIEW_EXTENSION_NAME, false);
	_register_requested_device_extension(VK_KHR_FRAGMENT_SHADING_RATE_EXTENSION_NAME, false);
	//_register_requested_device_extension(VK_KHR_CREATE_RENDERPASS_2_EXTENSION_NAME, false);
	_register_requested_device_extension(VK_KHR_SHADER_FLOAT16_INT8_EXTENSION_NAME, false);
```








![Bildschirmfoto 2024-04-25 um 17 53 05](https://github.com/godotengine/godot/assets/41921395/2a18c2f2-c26e-488f-bf70-669a37fd18a2)

DarioSamo commented:
@Alex2782 Can you confirm if https://github.com/godotengine/godot/pull/91169 fixes it? Thank you very much for the report by the way.
